aspnetcore/security/key-vault-configuration.md

```diff
@@ -12,11 +12,13 @@ uid: security/key-vault-configuration
 
 :::moniker range=">= aspnetcore-6.0"
 
-This article explains how to use the [Azure Key Vault](/azure/key-vault/) configuration provider to load app configuration values from Azure Key Vault secrets. Azure Key Vault is a cloud-based service that helps safeguard cryptographic keys and secrets used by apps and services. Common scenarios for using Azure Key Vault with ASP.NET Core apps include:
+This article explains how to use the [Azure Key Vault](https://azure.microsoft.com/services/key-vault/) configuration provider to load app configuration values from Azure Key Vault secrets. Azure Key Vault is a cloud-based service that helps safeguard cryptographic keys and secrets used by apps and services. Common scenarios for using Azure Key Vault with ASP.NET Core apps include:
 
 * Controlling access to sensitive configuration data.
 * Meeting the requirement for FIPS 140-2 Level 2 validated Hardware Security Modules (HSMs) when storing configuration data.
 
+[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/security/key-vault-configuration/samples) ([how to download](xref:index#how-to-download-a-sample))
+
```

Comment:
OK but I prefer the original where we don't duplicate the very verbose `view or download` before it's needed.

Fine to keep, but I prefer the original
<img width="434" alt="image" src="https://github.com/dotnet/AspNetCore.Docs/assets/3605364/c451a65f-79f9-4a5e-a034-03fbc804a760">

cc @tdykstra

Reply:
Most of the article is actually the same, too. Is there an easier way to vary just the code snippets for `< net6.0` vs. `>= net6.0`? There were a few other differences in a few paragraphs but I only briefly reviewed. I didn't want to boil the ocean, but a quick glance shows they really don't substantially vary by version either.

Reply:
> Most of the article is actually the same, too. Is there an easier way to vary just the code snippets for `< net6.0` vs. `>= net6.0`? There were a few other differences in a few paragraphs but I only briefly reviewed. I didn't want to boil the ocean, but a quick glance shows they really don't substantially vary by version either.

No. Very few customer are on .NET 5 and lower. .NET 6 and .NET 8 are LTS. It's very problematic to try to reduce duplication in sample code. It's prover easier to just make a copy and update it for the latest version.

Reply:
I'm fine with samples being separate. I mean, can I share most of the content in the page and only bifurcate the code snippets?

For example, would this work:

````markdown
This is common text for both pages.

:::moniker range=">= aspnetcore-6.0"
<!-- >= 6.0 sample code -->
:::moniker-end

:::moniker range="< aspnetcore-6.0"
<!-- < 5.0 sample code -->
:::moniker-end

More common text for both pages.
````

When I made two separate files from each section and diff'd them, there are almost no differences and those I briefly looked at probably don't need to be different. As I mentioned above: I just didn't want to boil the ocean. I could open a separate PR to make future edits easier, though. Not having to update duplicate text seems less error prone.

Reply:
> When I made two separate files from each section and diff'd them, there are almost no differences and those I briefly looked at probably don't need to be different

After maintaining an entire doc set for many versions we've found using multiple monikers in a file is hard to maintain and reason about. I'd say 90% of our docs have 97% duplication, and that has proven to be the most productive approach.

- Updating versions (where there are changes) is trivial and robust.
  - copy fileX.md to fileX/includes/filex7.md
  - remove meta data and change moniker from >= 7 to = 7
  - change moniker in fileX.md from >= 7 to >= 8
  - merge PR
  - create new PR with 8 changes. Now the diff file shows exactly what has changed.

We've found that mixing monikers:

- Is so difficult to reason about with professionals MSFT folks that mistakes are often made. 
- Is a significant hurdle for customers to update.

We've found out that:

- Most customers want the latest version. Older versions don't need fixes unless it's security related
- The eliminate duplication goal software folks embrace is highly problematic in docs.

Reply:
The file already uses multiple monikers, but those cover the entirety of the content. I would like to reduce maintenance costs by only bifurcating what needs to be different, which is pretty much just the code snippets.

Reply:
> The file already uses multiple monikers, but those cover the entirety of the content. I would like to reduce maintenance costs by only bifurcating what needs to be different, which is pretty much just the code snippets.

Right, that's because our moniker model has evolved. The current model has proven the most productive, that is moniker/version per include file.

>  I would like to reduce maintenance costs by only bifurcating what needs to be different

We've got enormous experience saying the opposite. Maybe you missed my point we don't update older versions unless theirs a security issue (which is rare).

Reply:
> We've got enormous experience saying the opposite. Maybe you missed my point we don't update older versions unless theirs a security issue (which is rare).

In this case, the text that disabled secrets were returned was wrong. You're suggesting I shouldn't update that older docs? I know .NET likes people to be on LTS, but even the Azure SDK targets `netstandard2.0` because a large percentage of customers still target .NET Framework. Having incorrect code even in older docs feels very wrong.

Again, I'm not suggesting trying to unify code snippets. I'm arguing that this:

````markdown
Common text.

:::moniker ">= aspnet-6.0"
newer code snippet
:::moniker-end

:::monker "< aspnet-6.0"
older code snippet
:::moniker-end

More common text
````

Seems better than this:

````markdown
:::moniker ">= aspnet-6.0"
Common text

newer code snippet

More common text
:::moniker-end

:::monker "< aspnet-6.0"
Common text

older code snippet

More common text
:::moniker-end
````

If we have to update "Common text", we do it only in one place vs. two. You're saying experience has shown the latter example is easier to maintain?

At least in the examples I'm seeing here, the only real reason we have different code snippets is because the configuration APIs were improved in latter versions and we can show more concise code.

Azure SDKs have a policy that the public API surface has to remain the same regardless of underlying support meaning, for example, we may have to throw a `NotSupportedException` against some targets or platforms, but the public API surface will still be there unless there's a good reason it can't be e.g., returns or accepts a new type that didn't previously exist - and even then, we try to abstract away those details if at all possible. So our "Common text" and "More common text" in the examples above will always be the same.